Update to new conway-polynomials python package

orlitzky opened the pull request:
Fix https://github.com/sagemath/sage/issues/32747 by switching to the newly-minted [conway-polynomials package](https://pypi.org/project/conway-polynomials/0.7/) on pypi.

IMO `sage.databases.conway` (which makes the dict immutable by wrapping it in a class) is of dubious value but I've left everything alone for now.

orlitzky committed:
build/pkgs/conway_polynomials: update to v0.7

orlitzky committed:
src/sage/databases/conway.py: update for conway-polynomials-0.7

orlitzky committed:
src/sage/features/databases.py: update for conway-polynomials-0.7

orlitzky committed:
src/sage/env.py: drop obsolete CONWAY_POLYNOMIALS_DATA_DIR

There's now a python package providing a standard interface to this
data.

mkoeppe commented on build/pkgs/conway_polynomials/SPKG.rst:
```diff
@@ -1,12 +1,26 @@
-conway_polynomials: Tables of Conway polynomials over finite fields
+Python interface to Frank Lübeck's Conway polynomial database
```
Need to keep `conway_polynomials:` at the beginning - that's the general format (for the TOC in the package list)

mkoeppe replied:
By the way, `./sage -package create conway_polynomials --pypi --type standard` sets up these things automatically in the right format

mkoeppe replied:
The `dependencies` file also needs updating

orlitzky replied:
Both should be fixed.

mkoeppe commented on src/sage/databases/conway.py:
```diff
@@ -152,7 +144,7 @@ def __len__(self):
             sage: len(c)
-            35352
+            35357
         """
```
Oh? What happened there?

kiwifb replied:
I was asking myself the same question.

orlitzky replied:
Inflation.

The old package came with preprocessed data that would have been harder to diff against the upstream list for changes. I think it was just a tiny bit out-of-date. Here's the diff:

```diff
103a104
> [2,110,[1,1,1,0,0,0,1,0,0,0,1,1,0,1,1,0,0,1,0,0,1,0,0,0,0,1,1,0,1,1,0,0,0,1,0,1,1,1,0,1,0,0,1,1,0,0,0,0,1,0,0,0,0,0,1,1,1,1,1,1,0,0,1,0,1,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]],
219a221
> [3,52,[2,0,1,1,1,2,0,1,2,0,0,0,1,0,2,0,2,2,2,1,2,0,2,1,0,0,2,2,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]],
222a225,226
> [3,56,[2,2,0,0,0,1,0,2,0,0,0,0,2,1,1,0,0,1,2,2,2,2,0,2,2,1,2,2,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]],
> [3,57,[1,1,2,2,2,2,1,0,1,1,0,2,1,0,2,2,2,1,0,2,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]],
229a234
> [3,72,[2,2,0,2,1,1,1,2,1,0,2,1,2,1,0,1,0,2,1,1,0,2,1,2,1,2,0,2,0,1,1,1,2,2,1,0,2,2,1,1,0,0,1,2,0,1,1,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]],
```

kiwifb replied:
I see. Does upstream update the list "often"?

orlitzky replied:
No idea. There's no change history on the site, but the new package ships the upstream file unmodified so at least we can `git diff` it in the future.

kiwifb replied:
Shipping upstream file unmodified sounds very good to me. I think it is ready.

orlitzky committed:
build/pkgs/conway_polynomials/SPKG.rst: fix header

Add back the "conway_polynomials: " prefix in the header to fix the
package's inclusion into the TOC.

orlitzky committed:
build/pkgs/conway_polynomials/dependencies: add $(PYTHON_TOOLCHAIN)

mkoeppe commented on build/pkgs/conway_polynomials/SPKG.rst:
```diff
@@ -1,4 +1,4 @@
-Python interface to Frank Lübeck's Conway polynomial database
+conway_polynomials: Python interface to Frank Lübeck's Conway polynomial database
 ===================================================================
```
```suggestion
=================================================================================
```

orlitzky replied:
argh

orlitzky committed:
src/sage/databases/all.py: Luebeck -> Lübeck to match conway.py

This fixes a failing doctest describing the database.

orlitzky committed:
src/sage/features/__init__.py: Luebeck -> Lübeck to match databases.py

This fixes a failing doctest for the description of the Conway
polynomials database.

mkoeppe reviewed (approved):
LGTM, thanks for working on this!

tobiasdiez commented on src/sage/features/databases.py:
```diff
@@ -44,16 +44,13 @@ def __init__(self):
-                            spkg='conway_polynomials',
-                            description="Frank Luebeck's database of Conway polynomials",
-                            type='standard')
+        PythonModule.__init__(
```
```suggestion
        super().__init__(
```
(not that it is that important)

tobiasdiez reviewed (changes requested):
Nice, thanks for working on this.

The conda tests are failing, e.g. because of
```
File "src/sage/combinat/designs/database.py", line 820, in sage.combinat.designs.database.OA_9_135
Failed example:
    OA = OA_9_135()                                                           # needs sage.rings.finite_rings sage.schemes
Exception raised:
    Traceback (most recent call last):
      File "/home/runner/work/sage/sage/src/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/runner/work/sage/sage/src/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.combinat.designs.database.OA_9_135[2]>", line 1, in <module>
        OA = OA_9_135()                                                           # needs sage.rings.finite_rings sage.schemes
      File "/home/runner/work/sage/sage/src/sage/combinat/designs/database.py", line 837, in OA_9_135
        G,B = singer_difference_set(16,2)
      File "/home/runner/work/sage/sage/src/sage/combinat/designs/difference_family.py", line 390, in singer_difference_set
        c = conway_polynomial(p,e*(d+1))
      File "/home/runner/work/sage/sage/src/sage/rings/finite_rings/conway_polynomials.py", line 63, in conway_polynomial
        return R(ConwayPolynomials()[p][n])
      File "/home/runner/work/sage/sage/src/sage/databases/conway.py", line 96, in __init__
        import conway_polynomials
    ModuleNotFoundError: No module named 'conway_polynomials'
```

So we need to also create a conda feedstock for it.

I think we can also remove the conway_polynomials feature (these failures show that it's presence or absence doesn't really imply anything).

orlitzky committed:
src/setup.cfg.m4: pull in conway_polynomial's install-requires as deps

conway_polynomials is a python package now and can be depended upon
via the python tooling.

orlitzky committed:
src/sage/features: forego detection of conway_polynomials

The conway_polynomials package is now a standard python dependency of
sagelib and does not need to be detected at runtime.

orlitzky committed:
src/sage/databases/conway.py: drop DatabaseConwayPolynomials import

The feature detection for Conway polynomials was removed; they're
always installed via a python package now.

mkoeppe commented on src/sage/features/databases.py:
```diff
@@ -19,43 +19,10 @@
     POLYTOPE_DATA_DIR)
 
 
-class DatabaseConwayPolynomials(StaticFile):
```
Normally I think we would say that this is API and should not be removed without deprecation. But I don't object to removing it

dimpase committed:
conda now has this package independently of Sage

tobiasdiez reviewed (approved):
LGTM

orlitzky committed:
build/pkgs/conway_polynomials: update to v0.8
